AI model server role

elikoga opened the pull request:
This PR introduces a new `ai-model-server` role for using Ollama for AI inference with acceleration on AMD GPUs.

PL-133949

@flyingcircusio/release-managers

## Release process

- [x] Created changelog entry using `./changelog.sh`

## PR release workflow (internal)

- [x] PR has internal ticket
- [x] internal issue ID (PL-…) part of branch name
- [x] internal issue ID mentioned in PR description text
- [x] ticket is on Platform agile board
- [ ] ticket state set to *Pull request ready*
- [ ] if ticket is more urgent than within the next few days, directly contact a member of the Platform team
<!---->
- [x] set urgency and risk labels
- [ ] ensure the merge bot has determined a merge date
- [ ] ensure all checks are green
- [ ] get a review from a colleague

## Design notes

- [ ] Provide a feature toggle if the change might need to be adjusted/reverted quickly depending on context. Consider whether the default should be `on` or `off`. Example: rate limiting.
- [ ] All customer-facing features and (NixOS) options need to be discoverable from documentation. Add or update relevant documentation such that hosted and guided customers can understand it as well.

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - This role provides an inference API which will handle potentially sensitive data. However, the ollama backend is stateless up to the KV cache.
  - ollama listens on the SRV interface which has the resource group firewall active to prevent unauthorised access to the inference backend. Access to ollama is handled by the AI gateway proxy.
- [x] Security requirements tested? (EVIDENCE)
  - Role manually verified on development inference server.

elikoga committed:
ai-model-server: add model server role for AI inference using Ollama

PL-133949

Co-authored-by: Molly Miller <mm@flyingcircus.io>

osnyx commented on nixos/roles/gpuserver.nix:
```diff
@@ -0,0 +1,176 @@
+  cfg = config.flyingcircus.roles.gpuserver;
+
+  # Pin specific package versions for model serving
+  pkgs_2505 = import (fetchTarball "https://releases.nixos.org/nixos/25.05/nixos-25.05.802491.7c43f080a7f2/nixexprs.tar.xz") { };
```
`builtins.fetchTarball` happens during eval phase and significantly slows down our Hydra evaluation loop. Please replace with one the fixed-output-derivation fetchers from `pkgs`, like `fetchzip`, which happen during build phase of their own derivation.

sysvinit replied:
Yes, this is a known issue, this PR is still a work in progress.

sysvinit committed:
ai-model-server: review fixes

Don't hardcode the ollama port, set the OLLAMA_HOST variable properly
to ensure the command line tools work properly, and remove unnecessary
firewall configuration otherwise covered by the standard RG firewall.

PL-133949

leona-ya commented:
We have decided on the name `ai-model-server`, that should be used here too.

leona-ya commented on nixos/roles/gpuserver.nix:
```diff
@@ -0,0 +1,170 @@
+      allowedTCPPorts = [
+        scfg.port
+      ];
+    };
```
This is most likely wrong. The port should be only open on SRV to members of the same resource group. This should already be achieved by `flyingcircus.firewall.enableSrvRgFirewall`.

sysvinit replied:
Good point.

sysvinit committed:
ai-model-server: add support for acceleration on amd gpus using rocm

Additionally simplify configuration of preloaded models

PL-133949

sysvinit committed:
ai-model-server: use upstream model preload functions

PL-133949

sysvinit committed:
ai-model-server: enable telegraf plugin for rocm

leona-ya commented on nixos/roles/ai-model-server.nix:
```diff
@@ -0,0 +1,157 @@
+          package = pkgs.ollama-rocm;
+          acceleration = "rocm";
+          # Pin to gfx1101 LLVM target for Radeon PRO W7900 GPUs
+          rocmOverrideGfx = "11.0.1";
```
Is this correct? https://rocm.docs.amd.com/en/latest/reference/gpu-arch-specs.html says gfx1100

sysvinit replied:
gfx1101 is what I blindly copied out of the hand-constructed ollama configuration, and it all seems to still work properly, but specifics of GPU configuration is a bit beyond me.

sysvinit committed:
ai-model-server: fixes from review feedback
